File: GS_FT/GS_FT.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 23 09:21:48 2019

@author: Siyuan Liu
"""
import numpy as np
import matplotlib.pyplot as plt
from numpy import fft
from numpy import pi

import skimage.io as io
from skimage import transform
import os




def gaussian1(res=64,w0=3e-3,L=12e-3):
    '''    
    res为分辨率，
    w0为光斑半径
    L为监视区域的边长
    单位m
    返回高斯光束电场强度
    '''
    x2=np.linspace(-L/2,L/2,res);                           #采样周期
    [X,Y]=np.meshgrid(x2,x2);
    r=np.sqrt(X**2+Y**2);
    E=np.exp(-r**2/w0**2).astype('complex128');                #电场高斯函数

    return E




res=64
target=io.imread('xiaolian.png')
target=transform.resize(target,(64,64),mode='constant')


iteration_time=100
E=gaussian1()
ang=np.random.randn(res,res)

for i in range(iteration_time):
    after_slm=np.exp(1j*ang)*E
    after_len=fft.fftshift(fft.fft2(after_slm))
    ang=np.angle(after_len)
    
    buffer=target*np.exp(1j*ang)
    before_len=fft.ifft2(fft.ifftshift(buffer))
    ang=np.angle(before_len)
    

ang=np.mod(ang/2/pi,1)
afterslm_end=np.exp(1j*ang*2*pi)*E
result=fft.fftshift(fft.fft2(afterslm_end))
nom_result=np.abs(result)/np.max(np.abs(result))

plt.figure(1)
plt.imshow(ang)

plt.figure(2)
plt.imshow(nom_result)

MSE=1/res*1/res*np.sum((nom_result-target)**2)
print('MSE= %f'%MSE)

io.imsave('aaa.tiff',ang)

File: GS_FT/GS_FT.m
clear;
clc;

target=im2double(imread('xiaolian.png'));

% target(40,54)=1;
% target(20,54)=1;
% target(54,10)=1;
inneriterationtime=100;

res=512;                   %分辨率
L=6e-3;                    %横向尺寸 单位m
%distance=0.2;                 %传播距离 单位m 取0，0.01，1，10
lambda=1.03e-6;             %波长 单位m
w0=3e-3;                    %光斑半径 单位m

% target=imresize(target,res/512)



x2=linspace(-L,L,res);      %横向尺寸 
dx=2*L/(res-1);             %采样周期
[X,Y]=meshgrid(x2,x2);
r=sqrt(X.^2+Y.^2);
E=exp(-r.^2/w0^2);          %电场高斯函数
I=exp(-2*r.^2/w0^2);        %光强高斯函数





ang=randn(res,res);
ang=zeros(res,res)

% afterdistance=abs(fun_fpropagation(gaussian_sqrt,H1)).^2;
% afterdistance=abs(afterdistance)/max(max(abs(afterdistance)));
% afterlen=fun_FT_Objective(afterdistance,1,lambda,focalLength);


for m = 1:inneriterationtime
    
    afterslm = exp(1i*ang).*E;
    afterlens1 =fun_FT_Objective(afterslm,1,lambda,0);
    temp = afterlens1;
    ang = angle(temp);

    buffer3=target.*exp(1i.*ang);
    beforelens1=fun_FT_Objective(buffer3,-1,lambda,0);
    ang = angle(beforelens1);
    
end

% ang=im2double(rgb2gray(imread('111.png')))*2*pi;
afterslm_end=exp(1i*ang).*E;
result=fun_FT_Objective(afterslm_end,1,lambda,0);
nom_result=abs(result)/max(max(abs(result)));
ang(ang<0)=ang(ang<0)+2*pi;

figure(1)
imshow(target);
figure(2)
imshow(nom_result);
figure(3)
imshow(abs(ang)/max(max(abs(ang))));


% imwrite(ang/2/pi,'holo_smile.png')

MSE=sum(sum((nom_result-target).^2))/64/64

% figure(1)
% imagesc(gaussian);
% figure(2)
% imagesc(abs(afterdistance));
